### PR TITLE
FEA-2422: Allow setting a default timeout for all requests

### PR DESCRIPTION
**Pull request by patkujawa-wf**
# [FEA-2422](https://jira.atl.workiva.net/browse/FEA-2422)
Issue Status

## Motivation
Having an unbounded request is rarely desirable and has caused issues with too many open, long-running connections.

## Changes
Set a (customizable) default timeout.

#### Release Notes
A new `defaultTimeoutThreshold` setter allows all requests without an explicit timeout to inherit that default.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Architecture member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/w_transport/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/w_transport/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/http/base_request.dart [hunk 1]
```diff
@@ -30,6 +30,9 @@ typedef ResponseInterceptor = Future<BaseResponse> Function(
     FinalizedRequest request, BaseResponse response,
     [RequestException error]);
 
+/// Specifies the default timeout for any request without an explicit one.
+Duration defaultTimeoutThreshold = Duration(minutes: 5);
```
Comment (patkujawa-wf) [inline lib/src/http/base_request.dart:34]: I suppose if it is public, we'll need to guard against null anyway.

File: lib/src/http/common/request.dart [hunk 2]
```diff
@@ -767,11 +767,11 @@ abstract class CommonRequest extends Object
...
 
-      // Enforce a timeout threshold if set.
-      Timer timeout;
-      if (timeoutThreshold != null) {
-        timeout = Timer(autoRetry.timeoutThreshold, _timeoutRequest);
+      // Enforce a timeout threshold.
+      if (timeoutThreshold == null) {
+        timeoutThreshold = defaultTimeoutThreshold;
       }
+      final timeout = Timer(autoRetry.timeoutThreshold, _timeoutRequest);
```
Comment (evanweible-wf) [inline lib/src/http/common/request.dart:774]: Someone could set `defaultTimeoutThreshold` to null, so we should still guard against that here.

**Comment by patkujawa-wf** [comment]
Boy, long tests. Skynet is just mad that GHA took longer than 10m 😆

**Comment by evanweible-wf** [comment]
@patkujawa-wf we're getting CI fixed up in this PR: https://github.com/Workiva/w_transport/pull/389

File: lib/src/http/common/request.dart [hunk 3]
```diff
@@ -766,11 +766,11 @@ abstract class CommonRequest extends Object
...
 
-      // Enforce a timeout threshold if set.
-      Timer? timeout;
-      if (timeoutThreshold != null) {
-        timeout = Timer(autoRetry.timeoutThreshold!, _timeoutRequest);
+      // Enforce a timeout threshold if specified.
+      if (timeoutThreshold == null && defaultTimeoutThreshold != null) {
+        timeoutThreshold = defaultTimeoutThreshold;
       }
+      final timeout = Timer(autoRetry.timeoutThreshold!, _timeoutRequest);
```
Comment (patkujawa-wf) [inline lib/src/http/common/request.dart:773]:
```suggestion
      final timeout = Timer(timeoutThreshold!, _timeoutRequest);
```
Might need to do that, actually. And what if `timeoutThreshold == null && defaultTimeoutThreshold == null`? Then the `!` will throw, right? So maybe the timeout does need to be nullable.

**Comment by patkujawa-wf** [comment]
Oh, I need to await the future in the test before the finally runs 🤦🏻

**Review by corwinsheahan-wf — APPROVED** [review]

**Review by evanweible-wf — APPROVED** [review]

**Comment by corwinsheahan-wf** [comment]
QA +1
* CI Passes (test added which cover changes)

Merging into master.
@Workiva/release-management-p

**Pull request merged**